### ISSUE TITLE
Is there some thing wrong in ConvOffset2D._to_bc_h_w_2 method?

### ISSUE THREAD

**Issue by mzhaoshuai** [felixlaumon/deform-conv#14]
Thx for your code!

I load a image and pass the image through a single `ConvOffset2D` layer, and save the output feature map.

The original image is like this(the size of the image is `1000x1000`):
![resie_image](https://user-images.githubusercontent.com/22476764/41148846-8b1b9910-6b3c-11e8-98e2-a9a5815b467e.jpg)

And the output feature map is like this:
![test_img_4_1](https://user-images.githubusercontent.com/22476764/41148903-b22db75e-6b3c-11e8-8146-e56f304e3546.jpg)

We can see the stucture of the image is broken.
So I modify the code of the `ConvOffset2D._to_bc_h_w_2` method,
```
@staticmethod
def _to_bc_h_w_2(x, x_shape):
	"""(b, h, w, 2c) -> (b*c, h, w, 2)"""
	x = tf.reshape(x, (x_shape[0], x_shape[1], x_shape[2], -1, 2))
	x = tf.transpose(x, [0, 3, 1, 2, 4])
	#x = tf.transpose(x, [0, 3, 1, 2])
	x = tf.reshape(x, (-1, int(x_shape[1]), int(x_shape[2]), 2))
	return x
```
I think the process of `[b, 2c, h, w] -> [bc, h, w, 2]`(directly) will not get the ideal results.

Re-run the code and I get the result
![test_img_4](https://user-images.githubusercontent.com/22476764/41149056-213909be-6b3d-11e8-9ce8-3308b28fe0df.jpg)
I think this maybe the right result...

The below code get similar result.
```
x = tf.transpose(x, [0, 3, 1, 2])
x = tf.reshape(x, (-1, 2, int(x_shape[1]), int(x_shape[2])))
x = tf.transpose(x, [0, 2, 3, 1])
```

The test code is like this:
```
def deform_test():
    ''' deformable cnn test'''
    image_holder = tf.placeholder(dtype = tf.float32, shape = [1000, 1000, 3])
    batch_image = tf.expand_dims(image_holder, axis = 0)
    offset_image = deform_layers.ConvOffset2D(batch_image, filters = 3 * 2,
                                                kernel_size = [3, 3],
                                                padding = 'same',
                                                name = 'dcnn')(batch_image)
    init_op = tf.global_variables_initializer()
    with tf.Session() as sess:
        sess.run(init_op)
        image_data = cv2.imread('img_4.jpg')
        image_data = cv2.resize(image_data, (1000, 1000))
        b_offset_image = sess.run([offset_image], feed_dict = {image_holder:image_data})
        cv2.imwrite('test_img_4.jpg', b_offset_image[0][0,...].astype('uint8')) 
```
I just use `tf.layers.Conv2D` to replace the keras function, the core code is same.

In fact, I am not total clear of the details of  data reshaping procedure. 
So I want to get some help of you.

**Comment by mzhaoshuai** [comment]
I do an additional experiment to test the reshape function

The code is like that:
```
import tensorflow as tf
import numpy as np


def shape_test():
	'''shape test '''
	def _to_bc_h_w_2_t1(x, x_shape):
		"""(b, h, w, 2c) -> (b*c, h, w, 2)"""
		x = tf.transpose(x, [0, 3, 1, 2])
		x = tf.reshape(x, (-1, int(x_shape[1]), int(x_shape[2]), 2))
		return x
	
	def _to_bc_h_w_2_t2(x, x_shape):
		"""(b, h, w, 2c) -> (b*c, h, w, 2)"""
		x = tf.reshape(x, (x_shape[0], x_shape[1], x_shape[2], -1, 2))
		x = tf.transpose(x, [0, 3, 1, 2, 4])
		x = tf.reshape(x, (-1, int(x_shape[1]), int(x_shape[2]), 2))
		return x

	def _to_bc_h_w_2_t3(x, x_shape):
		"""(b, h, w, 2c) -> (b*c, h, w, 2)"""	
		x = tf.transpose(x, [0, 3, 1, 2])
		x = tf.reshape(x, (-1, 2, int(x_shape[1]), int(x_shape[2])))
		x = tf.transpose(x, [0, 2, 3, 1])
		return x



	single_channel = np.array([[1,2], [4, 5]])
	feature_map = np.zeros((2,2,6))
	for i in range(feature_map.shape[-1]):
		feature_map[..., i] = single_channel

	feature_map = np.expand_dims(feature_map, axis = 0)
	print('original feature map:\n', feature_map, '\n')

	f_holder = tf.placeholder(dtype=tf.int32, shape=feature_map.shape)
	test_1 = _to_bc_h_w_2_t1(f_holder, f_holder.get_shape())
	test_2 = _to_bc_h_w_2_t2(f_holder, f_holder.get_shape())
	test_3 = _to_bc_h_w_2_t3(f_holder, f_holder.get_shape())

	with tf.Session() as sess:
		out_1, out_2, out_3 = sess.run([test_1, test_2, test_3], feed_dict={f_holder:feature_map})
		print('out_1:\n', out_1, '\n\n',
		 		'out_2:\n', out_2,'\n\n',
		 		'out_3:\n', out_3)


if __name__ == '__main__':
	shape_test()
```

The expected results should output the coordinate like `[1, 1]` or `[2, 2]` and so on。 

The result is :

```
original feature map:
 [[[[1. 1. 1. 1. 1. 1.]
   [2. 2. 2. 2. 2. 2.]]

  [[4. 4. 4. 4. 4. 4.]
   [5. 5. 5. 5. 5. 5.]]]]

out_1:
 [[[[1 2]
   [4 5]]

  [[1 2]
   [4 5]]]


 [[[1 2]
   [4 5]]

  [[1 2]
   [4 5]]]


 [[[1 2]
   [4 5]]

  [[1 2]
   [4 5]]]]

 out_2:
 [[[[1 1]
   [2 2]]

  [[4 4]
   [5 5]]]


 [[[1 1]
   [2 2]]

  [[4 4]
   [5 5]]]


 [[[1 1]
   [2 2]]

  [[4 4]
   [5 5]]]]

 out_3:
 [[[[1 1]
   [2 2]]

  [[4 4]
   [5 5]]]


 [[[1 1]
   [2 2]]

  [[4 4]
   [5 5]]]


 [[[1 1]
   [2 2]]

  [[4 4]
   [5 5]]]]

```
It looks like that the results of the original reshape method is wrong.

**Comment by mzhaoshuai** [comment]
I re-train the model with the new reshape method
(I choose this because `x = tf.reshape(x, (x_shape[0], x_shape[1], x_shape[2], -1, 2))` will cause exception. I guess maybe when `tensorflow` is constructing a static computation graph, the first dimension is `Nonetype` and the op is not proper  )
```
@staticmethod
def _to_bc_h_w_2(x, x_shape):
    """(b, h, w, 2c) -> (b*c, h, w, 2)"""
    x = tf.transpose(x, [0, 3, 1, 2])
    x = tf.reshape(x, (-1, 2, int(x_shape[1]), int(x_shape[2])))
    x = tf.transpose(x, [0, 2, 3, 1])
    return x
```
And the result is like

| Test Accuracy | Regular CNN | Deformable CNN |
| --- | --- | --- |
| Regular MNIST | 98.74% | 96.68% |
| Scaled MNIST | 65.96% | 96.91% |

It is just the final result, not the best. 

With the original reshape method 
```
    @staticmethod
    def _to_bc_h_w_2(x, x_shape):
        """(b, h, w, 2c) -> (b*c, h, w, 2)"""
        x = tf.transpose(x, [0, 3, 1, 2])
        x = tf.reshape(x, (-1, int(x_shape[1]), int(x_shape[2]), 2))
        return x
```
The result is like

| Test Accuracy | Regular CNN | Deformable CNN |
| --- | --- | --- |
| Regular MNIST | 99.14% | 86.95% |
| Scaled MNIST | 62.11% | 94.79% |

The best result of `Deformable CNN` on `Scaled MNIST` is  about `0.95`。

So I think the modified method got an obvious improvement. And I find that the runtime of the modified method is less than the original method.
